e2e/nomostest/clean.go

```diff
@@ -372,198 +313,207 @@ func resetSystemNamespaces(nt *NT, failOnError FailOnError) {
+	if err := nt.List(list, withLabelListOption(TestLabel, TestLabelValue)); err != nil {
+		if meta.IsNoMatchError(err) {
+			// Resource not registered. So none can exist.
+			return nil, nil
+		}
 		return nil, errors.Wrapf(err, "unable to clean objects of type %v from cluster", gvk)
 	}
-
 	return list.Items, nil
 }
 
-func deleteObject(nt *NT, u *unstructured.Unstructured) error {
-	finalizers := u.GetFinalizers()
-	if len(finalizers) == 1 && finalizers[0] == v1.SyncFinalizer {
-		// Special case logic for the SyncFinalizer, as objects may get stuck
-		// with it. We don't really care to wait for/rely on the controller to do
-		// its cleanup as it may have exited in an error state.
-		u.SetFinalizers([]string{})
-		err := nt.Client.Update(nt.Context, u)
-		switch {
-		case apierrors.IsNotFound(err) || meta.IsNoMatchError(err):
-			// The object isn't found, or the type no longer exists (in which case the
-			// object definitely doesn't exist and we're done.
-			return nil
-		case err != nil:
-			return errors.Wrapf(err, "unable to remove syncer finalizer from object %v: %s/%s",
-				u.GroupVersionKind(), u.GetNamespace(), u.GetName())
+func deleteObject(nt *NT, obj client.Object) error {
+	nn := client.ObjectKeyFromObject(obj)
+	gvk, err := kinds.Lookup(obj, nt.scheme)
+	if err != nil {
+		return err
+	}
+	if !obj.GetDeletionTimestamp().IsZero() {
+		// already terminating. skip deleting
+		return nil
+	}
+	if err = nt.Delete(obj, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return errors.Wrapf(err, "unable to delete %s object %s",
+				gvk.Kind, nn)
 		}
 	}
+	return nil
+}
 
-	err := nt.Client.Delete(nt.Context, u)
-	if err != nil && !apierrors.IsNotFound(err) {
-		return errors.Wrapf(err, "unable to clean test object from cluster %v: %s/%s",
-			u.GroupVersionKind(), u.GetNamespace(), u.GetName())
+func deleteObjects(nt *NT, objs ...client.Object) error {
+	for _, obj := range objs {
+		if err := deleteObject(nt, obj); err != nil {
+			return err
+		}
 	}
-
 	return nil
 }
 
+// deleteObjectsAndWait deletes zero or more objects in serial and waits for not found in parallel.
+// NOTE: Deleting in parallel might be faster, but the log would be harder to debug.
+func deleteObjectsAndWait(nt *NT, objs ...client.Object) error {
+	tg := taskgroup.New()
+	for _, obj := range objs {
+		nn := client.ObjectKeyFromObject(obj)
+		gvk, err := kinds.Lookup(obj, nt.scheme)
+		if err != nil {
+			return err
+		}
+		if !obj.GetDeletionTimestamp().IsZero() {
+			nt.T.Logf("[CLEANUP] already terminating %s object %s ...", gvk.Kind, nn)
+		} else {
+			nt.T.Logf("[CLEANUP] deleting %s object %s ...", gvk.Kind, nn)
+			if err := nt.Delete(obj, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+				if apierrors.IsNotFound(err) {
+					// skip waiting
+					continue
+				}
+				return errors.Wrapf(err, "unable to delete %s object %s",
```

Comment:
All of the clean deletion functions have this same problem. Some were trying to continue and some were exiting early. But trying to make it continue is pointless, because we don't want to run the tests if the test environment is dirty. If the environment is dirty we need to manually clean it up enough for Clean to finish the job.

So instead of continuing after some but not all of them, I made them all terminal. 

If Clean errors before the environment setup, it's fatal.
If Clean errors after the tests run, it's an error which will fail the suite after other registered Cleanups execute.

Reply:
The short version: Clean always needs to pass for a test to run. If Clean fails, it always runs again before tests can run. So Clean exiting early is fine.

Reply:
To be clear, I'm not saying to continue - just to wait for the already deleted resources to be NotFound. But perhaps it doesn't make much difference assuming we exit anyway after seeing this error

Reply:
I think I prefer exiting early/fast so we get the error faster and can triage faster. If the environment is dirty, it doesn't really matter how dirty.